### ISSUE TITLE
Feature List

### ISSUE THREAD

**Issue by smolinari** [skooppaos/appserver-io-cli#1]
Commands to be made available in the CLI.
- [x] "about" command. This is the default and explains how the CLI works.
- [x] "new" command. creates a new project.
  - Arguments/ Options:
    - [x] **name** - _mandatory_ - The name of the project.
    - [x] **[--with]** - _optional_ - the "with" parameter allows you to add different applications into the project automatically. For instance "--with routlt" will install the project with the `routlt`package.
  - Questions/ Parameters:
    - [x] If the name of the project is missing from command, ask for it.
    - [x] Ask for directory name.
    - [x] Ask for org name  
  - Tasks to do:
    - [x] Create project/ application directory.
    - [x] Create git repo.
    - [x] Create composer project.
    - [x] Create skeleton app.
    - [ ] If a package was added with ´--with´, install it.
- [ ] "install" command. To install full prebuilt applications.  "appserver install example" will install the example app automatically. 
- [ ] "restart" command. This is for restarting appserver. 
- [ ] "virtual-host" command. To create a new virtual host for a particular app. The app must be created first.
  - Arguments:
    **[name-of-directory]** - _mandatory_ - The name of the application directory.
  - .
- [ ] "remove" command. To remove an application, which is no longer needed. 
- [ ] "environment" command. Changes the environment.
  - Arguments:
    **[prod|dev]** - To select between production or development modes.
  - .
- [ ] "scanner" command. Creates a scanner to automatically restart appserver, when changes are made to a particular application's directory. 
  - Arguments:
    **[name-of-directory]** - _mandatory_ - The name of the directory under `/webapps`, which should be watched. 
  - Arguments: **[start|stop]** 

Scott

**Comment by wagnert** [comment]
Hi Scott, that's really awesome :)

Some things are not completely clear to me
1. Isn't it a bit confusing, when the `new` command with the `--with` Parameter allows you to install the example application. Wound't make it sense, to have `install` command that allow to install a application? Intentionally i would say, the `--with` Parameter should allow the user to add additional Composer libraries to the new project. What to you think?
2. A command to create a new Rout.Lt Action, Servlet or Bean is missing, correct?
3. I'm not sure, if a `mode` command makes sense, as this is already available by using `/opt/appserver/bin/appserver-runner`, but why not ;)

Beside that, i think this would be a really nice feature set for the start!

Tim

**Comment by smolinari** [comment]
Hey Tim. 

To 1.  Actually, I've been thinking about that too and I think I agree. I'll change "the plans" to include an "install" of the example app, instead of using `--with`. 

I am also still on the fence about "with" too. For one, because there is only the routlt extension to add. And for two, I feel if routlt is needed, it should actually be part of the core appserver and not part of the apps. Think of this. You have a blog, shop and forum apps. All three use routlt. Isn't that a bunch of extra code not really needed, if each app needs routlt? Or would the "apps" I just mentioned be under a single directory under `/webapps`? This is where I am still uncertain about how to build up a whole large scale application under appserver. The fact one can relate a virtual host within an app tells me, we could have say subdomains for each app. forum.my-site.com - my-site.com (for the blog) - shop.my-site.com, etc. Or am I standing on the hose again????

To 2. If that is something needed, then yes, let's do it! LOL!

To 3. Yeah, after rereading what the runner mode is all about, and hopefully understanding the docs, I guess you are right. I'll remove it.

Thanks for the feedback. 

Scott

**Comment by wagnert** [comment]
Hi Scott,

To 1. the `install` command is very appreciated :). As requested last week, i'll add `Rout.Lt` to the core, so that the last version will always be delivered with appserver.io. Additionally it should be possible to install a specific `Rout.Lt` version for each application, because there is a good chance that different applications (from different developers for example) uses different `Rout.Lt` versions and and the developer can not always know which version of appserver.io will be used by the system provider/hoster. Beside this, i think `--with` will be totally o. k., as it allows one to install additional composer libraries when creating a new project, e. g. someone will initially add the `JMSSerializer` library.

Usually each application will be in a separate directory unter `/webapps`, whereas each app can have as many subdomains as necessary, that's correct. 

To 2. I think for beginners it would be very helpful, so it would be a nice feature :)

You're work is very appreciated, so giving you feedback is the least i can do :)

Tim

**Comment by smolinari** [comment]
Talking about installing, I feel there is still two things missing. :smile_cat: 
1. A guide on how to properly extend appserver. Putting the routlt package into the core as an extension might be a cool example??? :smile: Also thanks for taking on that suggestion too. 
2. A guide to build apps to install in appserver. I know there is the deployment system and that is explained in the docs, but what needs to be in a deployment package for it to be a "proper" appserver app or extension isn't really explained. I'd actually like to duplicate the deploy system in the CLI too. (i.e. instead of messing with file extensions, just do an app deploy with `appserver install`.)

Scott

**Comment by wagnert** [comment]
Yes, you're right in that we've to extend our documentation for the topics :( Hm, it may also a nice idea to write a separate deployment in the CLI, i'll need some time to think about it and how it can be built :)
